### PR TITLE
Adjusts Real Sensor values,

### PR DESCRIPTION
**Pull request by dadamsretrorabbit**
Changes GetAlerts
Introduces Async for some Procedures.

### CODE REVIEW

File: WaterLog_Backend/Controllers/ReadingsController.cs [hunk 1]
```diff
@@ -61,8 +61,8 @@ public async Task Post([FromBody] InputSensor values)
             ReadingsEntry reading2 = new ReadingsEntry();
             reading.TimesStamp = DateTime.UtcNow;
             reading2.TimesStamp = DateTime.UtcNow;
-            reading.Value = values.valueIn;
-            reading2.Value = values.valueOut;
+            reading.Value = (values.valueIn)*60;
```
Comment (FredJEhlers) [inline WaterLog_Backend/Controllers/ReadingsController.cs]: 60 seems to be a constant please make use of a descriptive constant

File: WaterLog_Backend/Controllers/SegmentEventsController.cs [hunk 2]
```diff
@@ -82,22 +82,31 @@ public async Task AddSegmentEvent([FromBody] SegmentEventsEntry value)
...
         [Route("GetAlerts")]
         public async Task<List<GetAlerts>> GetAlerts()
+
         {
             try
             {
                 //Get all leaks first
-                var leaks = await _db.SegmentLeaks.Where(a => a.ResolvedStatus == EnumResolveStatus.UNRESOLVED).ToListAsync();
+                var leaks = await _db.SegmentLeaks.ToListAsync();
+                leaks = leaks.OrderByDescending(a => a.OriginalTimeStamp).OrderByDescending(a => a.ResolvedStatus).ToList();
```
Comment (FredJEhlers) [inline WaterLog_Backend/Controllers/SegmentEventsController.cs:91]: You have assumed you have gotten data from SegmentLeaks do a null check with appropriate error handling

File: WaterLog_Backend/Controllers/SegmentEventsController.cs [hunk 3]
```diff
@@ -82,22 +82,31 @@ public async Task AddSegmentEvent([FromBody] SegmentEventsEntry value)
...
                 if (leaks != null)
                 {
                     List<GetAlerts> alerts = new List<GetAlerts>();
                     var proc = new Procedures(_db, _config);
                     foreach (SegmentLeaksEntry entry in leaks)
                     {
+                        double totalSystemLitres = -1, litresUsed = -1,
```
Comment (FredJEhlers) [inline WaterLog_Backend/Controllers/SegmentEventsController.cs]: Declaring doubles but assigning ints. If you make use of a double your initialisation of these should be `-1.0`

File: WaterLog_Backend/Controllers/SegmentLeaksController.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ public async Task<ActionResult<string>> GetCost(int id)
                 return NotFound();
             }
             Procedures procedures = new Procedures(_db, _config);
-            return (JsonConvert.SerializeObject((procedures.CalculateTotalCost(leaks), procedures.CalculatePerHourWastageCost(leaks))));
+            return (JsonConvert.SerializeObject((procedures.CalculateTotalCostAsync(leaks), procedures.CalculatePerHourWastageCost(leaks))));
```
Comment (FredJEhlers) [inline WaterLog_Backend/Controllers/SegmentLeaksController.cs]: split over two lines

File: WaterLog_Backend/Controllers/SegmentLeaksController.cs [hunk 5]
```diff
@@ -88,9 +88,10 @@ public async Task<ActionResult<SegmentLeaksEntry>> Resolve([FromForm] int id)
 
         // GET api/segmentById/
         [HttpGet("{id}")]
-        public async Task<ActionResult<SegmentLeaksEntry>> Get(int id)
+        public async Task<ActionResult<SegmentLeaksEntry>> GetBySegmentId(int id)
         {
-            var leaks = await _db.SegmentLeaks.FindAsync(id);
+            //System makes assumption that segment and resolved status make entry distinct
+            var leaks = await _db.SegmentLeaks.Where(a => a.SegmentsId == id && a.ResolvedStatus == EnumResolveStatus.UNRESOLVED).FirstOrDefaultAsync();
```
Comment (FredJEhlers) [inline WaterLog_Backend/Controllers/SegmentLeaksController.cs]: split over two lines

File: WaterLog_Backend/Procedures.cs [hunk 6]
```diff
@@ -346,28 +349,38 @@ private string BuildUrl(int segmentId)
...
         }
 
-        public double CalculatePerHourWastageCost(SegmentLeaksEntry leak)
+        public async Task<double> CalculatePerHourWastageLitre(SegmentLeaksEntry leak)
         {
-            var list = _db.SegmentEvents;
-            var entry = list
-            .Where(inlist => inlist.SegmentsId == leak.SegmentsId)
-            .Last();
+            var list = await _db.SegmentEvents.Where(inlist => inlist.SegmentsId == leak.SegmentsId && inlist.EventType == "leak" && inlist.TimeStamp >= leak.OriginalTimeStamp && inlist.TimeStamp <= leak.LatestTimeStamp).ToListAsync();
```
Comment (FredJEhlers) [inline WaterLog_Backend/Procedures.cs]: split over multiple lines

File: WaterLog_Backend/Procedures.cs [hunk 7]
```diff
@@ -346,28 +349,38 @@ private string BuildUrl(int segmentId)
...
         }
 
-        public double CalculatePerHourWastageLitre(SegmentLeaksEntry leak)
+        public async Task<double> CalculatePerHourWastageCost(SegmentLeaksEntry leak)
         {
-            var list = _db.SegmentEvents;
-            var entry = list
-            .Where(inlist => inlist.SegmentsId == leak.SegmentsId)
-            .Last();
+            var list = await _db.SegmentEvents.Where(inlist => inlist.SegmentsId == leak.SegmentsId && inlist.EventType == "leak" && inlist.TimeStamp >= leak.OriginalTimeStamp && inlist.TimeStamp <= leak.LatestTimeStamp).ToListAsync();
```
Comment (FredJEhlers) [inline WaterLog_Backend/Procedures.cs]: split over multiple lines

File: WaterLog_Backend/Procedures.cs [hunk 8]
```diff
@@ -346,28 +349,38 @@ private string BuildUrl(int segmentId)
...
+            double usageperpoll = 0.0;
+            foreach(var item in list)
+            {
+                usageperpoll += (item.FlowIn - item.FlowOut);
+            }
+            //Convert Poll Into Minutes
+            usageperpoll *= 0.0167;
+            var minutes = (leak.LatestTimeStamp - leak.OriginalTimeStamp).TotalMinutes;
+            usageperpoll /= minutes;
+            usageperpoll *= 60.00;
```
Comment (FredJEhlers) [inline WaterLog_Backend/Procedures.cs]: Again 60 being used repeatedly. Make use of descriptive constants

File: WaterLog_Backend/web.config [hunk 9]
```diff
@@ -12,4 +12,4 @@
       <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
     </handlers>
   </system.webServer>
-</configuration>
+</configuration>
```
Comment (FredJEhlers) [inline WaterLog_Backend/web.config]: empty line EOF

File: WaterLog_Backend/Procedures.cs [hunk 10]
```diff
@@ -398,8 +419,9 @@ public async Task<double> CalculateTotalWastageLitres(SegmentLeaksEntry leak)
             .Where(inlist => inlist.SegmentsId == leak.SegmentsId)
             .Last();
 
-            var timebetween = (leak.LatestTimeStamp - leak.OriginalTimeStamp).TotalHours;
-            var perhour = CalculatePerHourWastageLitre(leak);
+            var timebetween = (leak.LatestTimeStamp - leak.OriginalTimeStamp).TotalMinutes;
+            timebetween /= 60.00;
+            var perhour = await CalculatePerHourWastageLitre(leak);
```
Comment (FredJEhlers) [inline WaterLog_Backend/Procedures.cs:430]: You are using this function quite a bit `CalculatePerHourWastageLitre` please make sure either that function have proper error handling to ensure that correct values are passed back, or check the values when it returns to ensure that you don't get unpredicted behaviour when you do calculations with the variables

**Review by FredJEhlers — CHANGES_REQUESTED** [review]

File: WaterLog_Backend/Controllers/SegmentEventsController.cs (hunk 2 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Controllers/SegmentEventsController.cs:91]: Code is encapsulated in a try and on line 92 there is a check for null.

File: WaterLog_Backend/Procedures.cs (hunk 10 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Procedures.cs:430]: We are using a Math.Max function to return 0 if any value falls into a negative.

File: WaterLog_Backend/Procedures.cs (hunk 8 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Procedures.cs]: Moved to Globals class

File: WaterLog_Backend/Procedures.cs (hunk 7 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Procedures.cs]: Done

File: WaterLog_Backend/Procedures.cs (hunk 6 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Procedures.cs]: Done

File: WaterLog_Backend/Controllers/SegmentLeaksController.cs (hunk 5 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Controllers/SegmentLeaksController.cs]: Done

File: WaterLog_Backend/Controllers/SegmentLeaksController.cs (hunk 4 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Controllers/SegmentLeaksController.cs]: Done

File: WaterLog_Backend/Controllers/SegmentEventsController.cs (hunk 3 above)
Comment (dadamsretrorabbit) [reply WaterLog_Backend/Controllers/SegmentEventsController.cs]: Done

**Review by FredJEhlers — APPROVED** [review]

**Pull request merged**